src/audio/google/google_rtc_audio_processing.c

```diff
@@ -639,6 +733,180 @@ static int google_rtc_audio_processing_reset(struct processing_module *mod)
+{
+#if CONFIG_COMP_GOOGLE_RTC_USE_32_BIT_FLOAT_API
+#if XCHAL_HAVE_HIFI3
+	const xtfloat ratio = 2 << 14;
+	xtfloat x0 = data;
+	xtfloat x1;
+	int16_t x;
+
+	x1 = XT_MUL_S(x0, ratio);
+	x = XT_TRUNC_S(x1, 0);
+
+	return x;
+#else /* XCHAL_HAVE_HIFI3 */
+	return Q_CONVERT_FLOAT(data, 15);
+#endif /* XCHAL_HAVE_HIFI3 */
+#else /* CONFIG_COMP_GOOGLE_RTC_USE_32_BIT_FLOAT_API */
+	return data;
+#endif
+}
+
+static inline BUF_TYPE convert_int16_to_google_aec_format(int16_t data)
+{
+#if CONFIG_COMP_GOOGLE_RTC_USE_32_BIT_FLOAT_API
+#if XCHAL_HAVE_HIFI3
+	const xtfloat ratio = 2 << 14;
+	xtfloat x0 = data;
+	float x;
+
+	x = XT_DIV_S(x0, ratio);
+
+	return x;
+#else /* XCHAL_HAVE_HIFI3 */
+	return Q_CONVERT_QTOF(data, 15);
+#endif /* XCHAL_HAVE_HIFI3 */
+#else /* CONFIG_COMP_GOOGLE_RTC_USE_32_BIT_FLOAT_API */
+	return data;
+#endif
+}
+
+/* todo CONFIG_FORMAT_S32LE */
+static int google_rtc_audio_processing_process(struct processing_module *mod,
+					       struct sof_source **sources, int num_of_sources,
+					       struct sof_sink **sinks, int num_of_sinks)
+{
+	int ret;
+	int16_t const *src;
+	int8_t const *src_buf_start;
+	int8_t const *src_buf_end;
+	size_t src_buf_size;
+
+	int16_t const *ref;
+	int8_t const *ref_buf_start;
+	int8_t const *ref_buf_end;
+	size_t ref_buf_size;
+
+	int16_t *dst;
+	int8_t *dst_buf_start;
+	int8_t *dst_buf_end;
+	size_t dst_buf_size;
+
+	size_t num_of_bytes_to_process;
+	size_t channel;
+
+	struct sof_source *ref_stream, *src_stream;
+	struct sof_sink *dst_stream;
+
+	struct google_rtc_audio_processing_comp_data *cd = module_get_private_data(mod);
+
+	if (cd->reconfigure) {
+		ret = google_rtc_audio_processing_reconfigure(mod);
+		if (ret)
+			return ret;
+	}
+
+	src_stream = sources[cd->raw_microphone_source];
+	ref_stream = sources[cd->aec_reference_source];
+	dst_stream = sinks[0];
+
+	num_of_bytes_to_process = cd->num_frames * source_get_frame_bytes(ref_stream);
+	ret = source_get_data(ref_stream, num_of_bytes_to_process, (const void **)&ref,
+			      (const void **)&ref_buf_start, &ref_buf_size);
+
+	/* problems here are extremely unlikely, as it has been checked that
+	 * the buffer contains enough data
+	 */
+	assert(!ret);
+	ref_buf_end = ref_buf_start + ref_buf_size;
+
+	/* de-interlace ref buffer, convert it to float */
+	for (int i = 0; i < cd->num_frames; i++) {
+		for (channel = 0; channel < cd->num_aec_reference_channels; ++channel) {
+			cd->aec_reference_buffer_ptrs[channel][i] =
+					convert_int16_to_google_aec_format(ref[channel]);
+		}
+		ref += cd->num_aec_reference_channels;
+		if ((void *)ref >= (void *)ref_buf_end)
+			ref = (void *)ref_buf_start;
+	}
+
+#if CONFIG_COMP_GOOGLE_RTC_USE_32_BIT_FLOAT_API
+	GoogleRtcAudioProcessingAnalyzeRender_float32(cd->state,
+						      (const float **)
+							cd->aec_reference_buffer_ptrs);
+#else
+	GoogleRtcAudioProcessingAnalyzeRender_int16(cd->state,
+						    (const int16_t *)cd->aec_reference_buffer);
```

Comment:
looks like `aec_reference_buffer` is not correctly populated properly for `int16` processing.

I would say: either we keep both option working (which complexify testing) or we choose only int16 or float32 (in that case my preference would go for the float32 one.